plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php

```diff
@@ -88,6 +129,455 @@ public function get_request_url( $order, $sandbox = false ) {
+			array(
+				'order_id'  => $order->get_id(),
+				'order_key' => $order->get_order_key(),
+				// Endpoint for the proxy to forward webhooks to.
+				'site_url'  => get_site_url(),
+				'site_id'   => class_exists( 'Jetpack_Options' ) ? Jetpack_Options::get_option( 'id' ) : null,
+			)
+		);
+
+		if ( strlen( $custom_id ) > 255 ) {
+			throw new Exception( 'PayPal order custom ID is too long. Max length is 255 chars.' );
+		}
+
+		return $custom_id;
+	}
+
+	/**
+	 * Get the order items for the PayPal create-order request.
+	 *
+	 * @param WC_Order $order Order object.
+	 * @return array
+	 */
+	private function get_paypal_order_items( $order ) {
+		$items = array();
+
+		foreach ( $order->get_items() as $item ) {
+			$items[] = array(
+				'name'        => $this->limit_length( $item->get_name(), self::PAYPAL_ORDER_ITEM_NAME_MAX_LENGTH ),
+				'quantity'    => $item->get_quantity(),
+				'unit_amount' => array(
+					'currency_code' => $order->get_currency(),
+					// Use the subtotal before discounts.
+					'value'         => wc_format_decimal(
+						$order->get_item_subtotal( $item, $include_tax = false, $rounding_enabled = false ),
+						wc_get_price_decimals()
+					),
+				),
+			);
+		}
+
+		return $items;
+	}
+
+	/**
+	 * Get the subtotal for all items, before discounts.
+	 *
+	 * @param WC_Order $order Order object.
+	 * @return float
+	 */
+	private function get_paypal_order_items_subtotal( $order ) {
+		$total = 0;
+		foreach ( $order->get_items() as $item ) {
+			$total += (float) $item->get_subtotal();
+		}
+
+		return $total;
+	}
+
+	/**
+	 * Get the value for the intent field in the create-order request.
+	 *
+	 * @return string
+	 */
+	private function get_paypal_order_intent() {
+		$payment_action = $this->gateway->get_option( 'paymentaction' );
+		if ( 'authorization' === $payment_action ) {
+			return 'AUTHORIZE';
+		}
+
+		return 'CAPTURE';
+	}
+
+	/**
+	 * Get the shipping preference for the PayPal create-order request.
+	 *
+	 * @param WC_Order $order Order object.
+	 * @return string
+	 */
+	private function get_paypal_shipping_preference( $order ) {
+		if ( ! $order->needs_shipping_address() ) {
+			return 'NO_SHIPPING';
+		}
+
+		$address_override = $this->gateway->get_option( 'address_override' ) === 'yes';
+		return $address_override ? 'SET_PROVIDED_ADDRESS' : 'GET_FROM_FILE';
+	}
+
+	/**
+	 * Get the shipping information for the PayPal create-order request.
+	 *
+	 * @param WC_Order $order Order object.
+	 * @return array
+	 */
+	private function get_paypal_order_shipping( $order ) {
+		if ( ! $order->needs_shipping_address() ) {
+			return null;
+		}
+
+		$address_type = 'yes' === $this->gateway->get_option( 'send_shipping' ) ? 'shipping' : 'billing';
+
+		return array(
+			'name'    => array(
+				'full_name' => $order->{"get_formatted_{$address_type}_full_name"}(),
+			),
+			'address' => array(
+				'address_line_1' => $this->limit_length( $order->{"get_{$address_type}_address_1"}(), 300 ),
```

Comment:
I would consider using constants for the length limit values as well. But that's also optional (as everything else)